### DISCUSSION TITLE
How to use `Arc<dyn PhysicalExpr>` inside `HashMap` as key, or inside `HashSet`.

### DISCUSSION THREAD

**Discussion by mustafasrepo** [apache/datafusion#7906] [General]
I think some of the mappings during projection or in other parts of the codebase, can be simplified. If we were to use `Arc<dyn PhysicalExpr>` inside a `HashMap` (e.g `HashMap<Arc<dyn PhysicalExpr>, Arc<dyn PhysicalExpr>>`). However, I gave it a try, and couldn't do so. 

Is there a way to do so? I have tried to use a wrapper struct inside `HashMap` such as `HashMap<ExprWrapper, ExprWrapper>`
where `ExprWrapper` is 
```rust
struct ExprWrapper{
    expr: Arc<dyn PhysicalExpr>,
}

impl PartialEq for ExprWrapper {
    fn eq(&self, other: &ExprWrapper) -> bool {
        self.expr.eq(&other.expr)
    }
}

impl Eq for ExprWrapper {}

impl Hash for ExprWrapper {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.expr.hash(state);
    }
}
```
This suggests me that, there should be a way to use directly `Arc<dyn PhysicalExpr>>` inside `HashMap`. I wonder if there is anyone who has done similar to this before, and How can I do so?

Thanks in advance.

**Comment by alamb** [comment]
What is the problem with using `HashMap<Arc<dyn PhysicalExpr>, Arc<dyn PhysicalExpr>>`

We implemented `Hash` for `PhysicalExpr` for just the purpose I thought 🤔 

https://docs.rs/datafusion/latest/datafusion/physical_expr/trait.PhysicalExpr.html#impl-Hash-for-dyn+PhysicalExpr

**Reply by mustafasrepo** [reply]
Actually we can create the hashMap for the type `HashMap<Arc<dyn PhysicalExpr>, Arc<dyn PhysicalExpr>>`. Below code works correctly.
```rust
    // Ok, no error
    #[test]
    fn test_hash_map_creation() -> Result<()> {
        let col_a = &Column::new("a", 0);
        let col_b = &Column::new("b", 1);
        let expr_a = Arc::new(col_a.clone()) as Arc<dyn PhysicalExpr>;
        let expr_b = Arc::new(col_b.clone()) as Arc<dyn PhysicalExpr>;
        let hash_map: HashMap<Arc<dyn PhysicalExpr>, Arc<dyn PhysicalExpr>> = HashMap::new();
        Ok(())
    }
```
However, when we want to insert something to the hashmap it gives an error. Code below produces error the method `insert` exists for struct `HashMap<Arc<dyn PhysicalExpr>, Arc<dyn PhysicalExpr>>`, but its trait bounds were not satisfied.
```rust
    // Gives the error "the method `insert` exists for struct `HashMap<Arc<dyn PhysicalExpr>, Arc<dyn PhysicalExpr>>`, but its trait bounds were not satisfied";
    #[test]
    fn test_hash_map_insertion() -> Result<()> {
        let col_a = &Column::new("a", 0);
        let col_b = &Column::new("b", 1);
        let expr_a = Arc::new(col_a.clone()) as Arc<dyn PhysicalExpr>;
        let expr_b = Arc::new(col_b.clone()) as Arc<dyn PhysicalExpr>;
        let mut hash_map: HashMap<Arc<dyn PhysicalExpr>, Arc<dyn PhysicalExpr>> = HashMap::new();
        hash_map.insert(expr_a, expr_b);
        Ok(())
    }
```

**Comment by mustafasrepo** [comment]
It seems that when we use a type with unknown size inside `Arc`. `Eq` property is not satisfied. This prevents us to use this type as key in the `HashMap`. However, I am surprised this error doesn't arise during initialization
